Add topic replication factor change (#101)
strimzi/proposals: 063-topic-replication-factor-change.md

When data loss is not a problem, the user can simply delete and recreate the KafkaTopic resource with the desired replication factor.
This is not really an option for most use cases.

## Motivation

The topic replication factor change feature will provide a much better cloud-native experience for Strimzi users.
They will be able to change the topic replication factor by simply updating the KafkaTopic resource, which may also happen through a GitOps pipeline.

This is a non exhaustive list of use cases that will become significantly easier:

- Set a lower replication factor for non-critical topics to reduce the infrastructure costs (compute, storage and network).
- Set a lower replication factor to deal with resource shortage (trade-off between fault-tolerance and resource utilization).
- Set a higher replication factor for critical topics to improve durability and fault tolerance.

## Proposal

The replication factor change feature can be implemented by integrating the Topic Operator with Cruise Control.
The Cluster Operator is already integrated with Cruise Control, so we can factor out the existing logic and classes.

### Cruise Control

> Cruise Control creates a task for any operation, such as topic configurations and cluster rebalances.

Reviewer: What _operation_ is _topic configurations_?

< Content-Length: 17051
< Server: Jetty(9.4.52.v20230823)
< 
{ [17321 bytes data]
{
  "summary": {
    "numIntraBrokerReplicaMovements": 0,
    "numReplicaMovements": 125,
    "onDemandBalancednessScoreAfter": 100,
    "intraBrokerDataToMoveMB": 0,
    "monitoredPartitionsPercentage": 100,
    "provisionRecommendation": "",
    "excludedBrokersForReplicaMove": [],
    "excludedBrokersForLeadership": [],
    "provisionStatus": "RIGHT_SIZED",
    "onDemandBalancednessScoreBefore": 100,
    "recentWindows": 5,
    "dataToMoveMB": 0,
    "excludedTopics": [],
    "numLeaderMovements": 0
  }
  ...
```

The `goals` parameter is the comma separated list of goals used to generate the configuration proposal.
This is configurable for all endpoints in `Kafka.spec.cruiseControl.config` with `default.goals` property.

The `replication_throttle` parameter is an upper bound on the bandwidth used to move replicas (bytes per second).
This is configurable for all endpoints in `Kafka.spec.cruiseControl.config` with `default.replication.throttle` property.

The `skip_rack_awareness_check` boolean parameter is undocumented and disabled by default.

> In case of a single-rack deployment, rack goals are not configured and the Topic Operator will automatically enable `skip_rack_awareness_check`.

Reviewer: So, are these things known to Cruise Control? Or are they passed through the request by the Topic Operator?

Author: They are passed through the request by the Topic Operator. It's the Cluster Operator that sets STRIMZI_CRUISE_CONTROL_RACK_ENABLED env var, or the user in case of standalone deployment.

Reviewer: Maybe worth adding here a sentence or two that these parameters will need to be passed to the TO since the list of env vars.

Reviewer: Actually, the `STRIMZI_CRUISE_CONTROL_RACK_ENABLED` is mentioned int he environment variables. But nothing about goals or replication throttle.

Author: They are not configurable. I'm mentioning them in the "Topic configuration endpoint" section to say that they can be configured from Kafka.spec.cruiseControl.config.

Reviewer: Ok. Maybe we should clarify it in the proposal as it is a bit confusing.

```sh
$ curl -s "localhost:9090/kafkacruisecontrol/user_tasks?user_task_ids=5344ca89-351f-4565-8d0f-9aade00e053d&json=true" | jq
{
  "userTasks": [
    {
      "Status": "Completed",
      "ClientIdentity": "127.0.0.1",
      "RequestURL": "POST /kafkacruisecontrol/topic_configuration?topic=my-topic&replication_factor=2&skip_rack_awareness_check=true&dryrun=false&json=true",
      "UserTaskId": "5344ca89-351f-4565-8d0f-9aade00e053d",
      "StartMs": "1701961135978"
    }
  ],
  "version": 1
}
```

### Topic Operator

At runtime, the Topic Operator watches and periodically reconciles the configuration of all managed and unpaused KafkaTopic resources.
Among the other configurations, the reconciliation logic also detects `.spec.replicas` changes by comparing with the actual topic replication factor in Kafka.
From the point of view of the Topic Operator, a topic replication factor change can be in the following states:

- **Pending**: replicas is updated, but the replication factor change request is not sent to Cruise Control.
- **Ongoing**: replicas is updated and Cruise Control has accepted the request, but the execution is not completed.

> - **Ready**: replicas is updated and Cruise Control execution is completed.

Reviewer: What do these correspnd to in terms of the Cruise Cotnrol states? Is Pending a task that is Active but not in execution yet? Or is Pending the state when the CC queue is full and the tasks is not queued yet? I guess the later?

Author: > Is Pending a task that is Active but not in execution yet?

No, that's the Ongoing state. Or better, an ongoing change is when the request has been accepted by CC (it's in the queue). In terms of CC states, it can be Active or InExecution. Proposal updated.

> is Pending the state when the CC queue is full and the tasks is not queued yet?

In that case, the Topic Operator gets an `OngoingExecutionException`, updates the KafkaTopic status with an appropriate message, waits for some time (backoff period), and then removes the KT from pending changes. The last operation allows a new notification from the periodic reconciliation to come in, which is basically a retry.

I added this example to the proposal.

When Cruise Control integration is enabled and a replicas change is detected by the BatchingTopicController, the event is notified to the ReplicasChangeManager.

> The ReplicasChangeManager is responsible for managing the communication with Cruise Control's REST API and updating the KafkaTopic status.

Reviewer: You should probably first introduce the `ReplicationChangeManager` and explain what it is and how it works.

Author: I changed the words a bit. Let me know if it's enough.

> <p align="center"><img src="images/063-topic-replication-factor-change.png"/></p>

Reviewer: ```suggestion
![Replication Factor Change](./images/063-topic-replication-factor-change.png)
```

The KafkaTopic resources with replicas change are added to a pendingChanges list the first time they are notified to the ReplicasChangeManager.

> When a replicas change is marked as pending, other replicas changes for the same KafkaTopic are ignored until the pending replicas change is sent to Cruise Control and the execution is completed.

Reviewer: * How would this be implemented?
* Why would this be the case if the Pending state means it is not queued in CC yet? It seems in such case you can easily accept new changes or just cancel it.

Author: The periodic reconciliation logic continues to detect and notify the replicas change to the ReplicasChangeManager. After the first, the ReplicasChangeManager ignores other replicas change notifications for the same topic using the pendingChanges cache. Once the current pending is done, the new change becomes pending and can be executed.

This has the benefit of ignoring intermediate changes in case they are done in a relatively short period of time, and only apply the first and last change, also considering that CC runs one task at a time, and the queue is shared with all other tasks (i.e. rebalances).

When a new replicas change for the same topic comes in, we may consider stopping the current replicas change by removing it from pendingChanges cache, and doing a stop request to CC in case it is also in ongoingChanges. I didn't do that, because it would be a lot of work and wasted cycles for CC. When you stop a task, the current batch still completes, and it could take some time.

Reviewer: It seems that this would not survive restarts of the operator. Also, there seems to be no need for this. At least until the task is queued within CC (that seems to include the pending state mentioned here) there is no reason to not reflect the latest change.

Author: > It seems that this would not survive restarts of the operator. 

Pending changes that are also ongoing, will be restored on ReplicasChangeManager initialization when the TO restarts. The rest of pending changes will be added as if they were new as soon as the first periodic reconciliation runs.

> Also, there seems to be no need for this. At least until the task is queued within CC (that seems to include the pending state mentioned here) there is no reason to not reflect the latest change.

The change is reflected in KafkaTopics, but with a Condition that says ongoing, so I think it is correct.

Reviewer: I still don't think this is recoverable. The condition does not store the number of replicas. So if the operator restarts, you cannot recover from it what the change was. I also think recovering from the conditions is weird. I haven't seen that done anywhere. I would not use conditions to store information but only to inform users.

The KafkaTopic resources marked as pending are added to an ongoingChanges map keyed by `User-Task-ID` when the topic configuration request is accepted by Cruise Control.

> This map is used to periodically check Cruise Control's tasks and update the KafkaTopic status.

Reviewer: What kind of map are we talking here? Config Map? Some Java in memory map?

Author: Yes, it's a Java in-memory map.

|:---------------------------------------|---------|---------|---------------------------------------------------------------|
| STRIMZI_CRUISE_CONTROL_ENABLED         | Boolean | false   | Whether Cruise Control integration is enabled                 |
| STRIMZI_CRUISE_CONTROL_NAMESPACE       | String  | ""      | Namespace where Cruise Control is running                     |
| STRIMZI_CRUISE_CONTROL_HOSTNAME        | String  | ""      | Cruise Control hostname                                       |
| STRIMZI_CRUISE_CONTROL_PORT            | Integer | 9090    | Cruise Control port                                           |
| STRIMZI_CRUISE_CONTROL_SSL_ENABLED     | Boolean | false   | Whether SSL encryption is enabled                             |
| STRIMZI_CRUISE_CONTROL_AUTH_ENABLED    | Boolean | false   | Whether Basic authentication is enabled                       |
| STRIMZI_CRUISE_CONTROL_RACK_ENABLED    | Boolean | false   | Whether rack awareness is enabled in the target Kafka cluster |
| STRIMZI_CRUISE_CONTROL_SECRET_NAME     | String  | ""      | Secret name containing the truststore                         |

> | STRIMZI_CRUISE_CONTROL_API_SECRET_NAME | String  | ""      | Secret name containing the API credentials                    |

Reviewer: Please explain why not mount these directly (if it has any advantages). Mounting them will make a lot of things much easier:
* RBAC
* Not overloading Kube APIs
* Expecially in the standalone operator case also the structure of the secrets.

You should also explain what credentials will it use and how will they be managed.

Author: > Please explain why not mount these directly (if it has any advantages).

Do you mean using the env vars to pass truststore and credentials data, instead of secret names right? One problem I see is that CC credentials are generated after the TO is deployed, so I wouldn't know what to mount in the Cluster Operator deployment use case. Then, I would need 4 env vars instead of 2.

> You should also explain what credentials will it use and how will they be managed.

This is explained in the configuration section.

Author: @scholzj if it's fine to deploy CC first and then the EO when CC integration is enabled, then I can explore this solution.

Reviewer: To be clear, it is not something I tried. Maybe I'm missing some interdependency. But it seems to me like something what should work and migth be the easiest way to address this.

Author: I'm adopting your suggestions here, and removing the Secret RBAC from the draft implementation. See config parameters section. Thanks.

In case of Cluster Operator deployment, a dedicated admin user is automatically created for the Topic Operator.

In case of standalone deployment, the Cluster Operator expects the following secrets to be created by the user:

- A secret with `cruise-control.p12` key containing the truststore in PKCS12 format, and `cruise-control.password` key containing the truststore password.
- A secret with `cruise-control.apiToAdminUsername` key containing the API username with admin role, and `cruise-control.apiToAdminPassword` key containing the user password.

#### Error handling

A replicas change is indefinitely retried by the periodic reconciliation when Cruise Control's task queue is full, or there is an error.

> The user is always informed about the replicas change state through appropriate logs and KafkaTopic status updates.

Reviewer: So, how do you cancel or recover the change? Earlier, you suggested that once the RF is changed, any further changes are ignored. But if I by mistake change the the RF to 4 on my 3 node cluster, the change will be presumably failed by CC and I would want to cancel it (e.g. by changing the RF to some valid number). How would that work?

Author: All changes with errors, or completed with errors are automatically removed from pendingChanges and the user will be able to fix the error by updating the KT. If they don't do it, the TO will continue to retry (after some backoff period to not create too much work for CC) and log the same error.

Author: All changes with errors, or completed with errors are automatically removed from pendingChanges and the user will be able to fix the error by updating the KT. If they don't do it, the TO will continue to retry (after some backoff period to not create too much work for CC) and log the same error.

During the whole replicas change process, the KafkaTopic resource remains in a Ready state.

> Any KafkaTopic with replicas change has a status condition of type Ongoing and reason ReplicasChanged, which contains the CruiseControl `User-Task-ID` in the message.

Reviewer: Seems weird to hide this in status condition. I think oyu should consider a separate subsection in the KafkaTopic CR status section. (For the Task ID => condition is fine to inform about issues and progress etc.)

Author: Ok, that's a good idea. I was just waiting for someone to raise it :)

This happens because Cruise Control has no persistent memory of active tasks.
In this case, the Topic Operator resets the state related to that `User-Task-ID`, which unblocks notifications of all not executed replicas changes.

Cruise Control allows to scale down the replication factor under the `min.insync.replicas` value, and this can cause an availability issue for producers with `acks=all`.
When this happens, the Topic Operator logs a warning and continues with the change because the KafkaRoller ignores topics with RF < minISR, as they don't event show up as under replicated.
The target replication factor should also be less than or equal to the number of available brokers, but this is enforced directly by Cruise Control.

When multiple changes to the same KafkaTopic resource arrive in a short period of time, they are all ignored except the first and the last changes.
When a manged topic with replicas change is deleted in Kubernetes, the Topic Operator also deletes it in Kafka, but the ongoing task execution in Cruise Control will continue logging an error.

> When a manged topic with replicas change is deleted in Kafka, the Topic Operator recreates the topic with the target replication factor.

Reviewer: Shouldn't you cancel it? Or how will it be cleared from Cruise Control?

> POST /kafkacruisecontrol/topic_configuration?skip_rack_awareness_check=true&dryrun=false&json=true HTTP/1.1
> Host: localhost:9090
> User-Agent: curl/8.0.1
> Accept: */*
> 
} [91 bytes data]
< HTTP/1.1 200 OK
< Date: Thu, 07 Dec 2023 14:58:55 GMT
< Set-Cookie: JSESSIONID=node0lfkx183ijq4pjkkjahsegfmp30.node0; Path=/
< Expires: Thu, 01 Jan 1970 00:00:00 GMT
< User-Task-ID: 5344ca89-351f-4565-8d0f-9aade00e053d
< Content-Type: application/json;charset=utf-8
< Cruise-Control-Version: 2.5.77-SNAPSHOT
< Cruise-Control-Commit_Id: 535ea07e8847990cc02dde1f1be99b387dbeaf5b
< Content-Length: 17051
< Server: Jetty(9.4.52.v20230823)
< 
{ [17321 bytes data]
{
  "summary": {
    "numIntraBrokerReplicaMovements": 0,
    "numReplicaMovements": 125,
    "onDemandBalancednessScoreAfter": 100,
    "intraBrokerDataToMoveMB": 0,
    "monitoredPartitionsPercentage": 100,
    "provisionRecommendation": "",
    "excludedBrokersForReplicaMove": [],
    "excludedBrokersForLeadership": [],
    "provisionStatus": "RIGHT_SIZED",
    "onDemandBalancednessScoreBefore": 100,
    "recentWindows": 5,
    "dataToMoveMB": 0,
    "excludedTopics": [],
    "numLeaderMovements": 0
  }
  ...
```

> The `goals` parameter is the comma separated list of goals used to generate the configuration proposal.

Reviewer: You mention the rebalance proposal here. But do not cover it later. Will it be used by TO? Or will TO directly change the RF? It would be worth explaining.

Author: TO requests the RF change to CC. Quoted from the proposal: "Under the cover, Cruise Control reuse the same heuristic used for the rebalance operation to determine the best location to place or remove partition replicas."

Reviewer: Sorry, maybe my comment here wasn't clear. What the sentence in the proposal suggests to me is that CC gives you a proposal of how it will add the replicas and you approve it. Is that the case or not?

Author: No, it's all automatic. I'll add a clarification.

      "UserTaskId": "5344ca89-351f-4565-8d0f-9aade00e053d",
      "StartMs": "1701961135978"
    }
  ],
  "version": 1
}
```

### Topic Operator

At runtime, the Topic Operator watches and periodically reconciles the configuration of all managed and unpaused KafkaTopic resources.
Among the other configurations, the reconciliation logic also detects `.spec.replicas` changes by comparing with the actual topic replication factor in Kafka.
From the point of view of the Topic Operator, a topic replication factor change can be in the following states:

- **Pending**: replicas is updated, but the replication factor change request is not sent to or accepted by Cruise Control.
- **Ongoing**: replicas is updated and Cruise Control has accepted the request (task queued), but the execution is not completed (possible task states: Active and InProgress).
- **Ready**: replicas is updated and Cruise Control task execution is completed (possible task states: Completed).

When Cruise Control integration is enabled and a replicas change is detected by the BatchingTopicController, the event is notified to the ReplicasChangeManager.

> The ReplicasChangeManager maintains a cache of pending and ongoing replicas changes, manages the communication with Cruise Control's REST API and updates the KafkaTopic status.

Reviewer: You keep ignoring my comments and questions on this one. I'm missing an explanation why this is needed - the `ReplicasChangeManager` and its cache ... or how will it work. My assumption is that the easiest way to handle this is in the regular reconciliation:
1. You check the status if there is already a CC task (and either handle it if is it complete and remove it from status or just check it and keep going if it is still queued or in progress)
2. If there is no task in the status, you check if the replicas in Kafka and in the KafkaTopic resource are the same. And if not, you _try to create the task in CC_.

What do you need the `ReplicasChangeManager` and its cache for and how will it work? Maybe there is some reason, but it seems you are not really explaining it anywhere and just closing my comments.

Author: ### Isolate complexity

It is always the regular reconciliation that drives the ReplicasChangeManager, by providing replicas change events. The Cruise Control REST API integration is quite complex with different HTTP requests and polling. Using a dedicated object seems to be a natural choice to isolate that complexity from the rest of the reconciliation logic. This is somewhat similar to the choice of providing a Reconciler object where a single component reconciliation logic is hosted.

### Decouple asynchronous reconciliation

A replicas change reconciliation is very different from the reconciliation of other topic configurations, as it is executed asynchronously. After a replicas change event is notified, it takes some time for the request to land in Cruise Control's queue, and some more time to be actually executed. Most reconciliations will find that the replicas change is still pending or ongoing, due to the limited Cruise Control capacity (only one task can be executed at a time). This would generate a lot of unnecessary HTTP requests, that would slowdown the reconciliation logic for no good. 

### Avoid duplicates and enable batch mode

The ReplicasChangeManager's cache is needed to avoid generating duplicated requests that would overload Cruise Control and fill the task queue with unnecessary tasks, with the effect of delaying valid replicas change requests. It is also used to group multiple replicas changes together and map them to a single Cruise Control request/task (batch mode). This is useful in case if bulk replicas change (i.e. gitops update).

Let me know if it is clear. 
Will add these points to the proposal.

Reviewer: Thanks for providing the reasoning. While I might not agree with it in all cases, I think it is definitely useful to have it in the proposal as an explanation. That said, this does not explain anything about how will the `ReplicasChangeManager` work. All I have are some guesses based on some parts of it you mention such as the cache etc. So it would be good to cover that as well.

----

> Isolate complexity

I do not think nobody prevents you from having classes to abstract the CC API Cruise Control complexity. But I do not think you provide any reasons why to have a _separate operator_ for it. The polling can be easily done from the regular reconciliations. Or maybe you badly explain how will the `ReplicasChangeManager` work.

The Reconciler classes in the Cluster Operator - if that is what you were referring to - are stateless. They do not keep any state (cache) and are called only from resource reconciliation. I think that is a valid pattern but your reasoning does not seem to suggest this is the same as what you are planing.

> Decouple asynchronous reconciliation

I do not understand the promised _efficiency. The reconciliation of a KafkaTopic CR happens normally every 2? minutes (Or does TO still use 90 seconds?) and on changes. That does not seem to overload any CC APIs. How often will the `ReplicasChangeManager` call it? Or how would it reduce the API calls?

> Avoid duplicates and enable batch mode

Duplicates can be avoided when driving this through the custom resources as well => at the end this is the backup for operator restarts. You either have the task ID you got stored in it or not. Batching in my mind does not justify the complexity. Do you really expect bulk changes in the replication factor to be a common thing? What would be the use-cases? _Latency_ of how quickly the changes are applied also does not seem to be a major issue here.

Author: Let me answer some questions, then I'll try to add more details to the proposal.

> The reconciliation of a KafkaTopic CR happens normally every 2? minutes (Or does TO still use 90 seconds?) and on changes. That does not seem to overload any CC APIs. 

The backlog of changes may be big not only after a bulk update, but also when CC is running a long rebalance task. In this case, all replicas changes have to wait and accumulate in the TO. Assuming no batching, this would generate 2 HTTP requests for each replicas change on every periodic reconciliation.

> How often will the ReplicasChangeManager call it? Or how would it reduce the API calls?

The ReplicasChangeManager calls the configuration endpoint just one time, with all unknown replicas changes. If there a no new replicas changes, no request is sent. Then, it polls the tasks endpoint every 10 seconds passing all ongoing changes (one request). In case of any error, it will wait for 60 seconds and then retry, indefinitely.

> Duplicates can be avoided when driving this through the custom resources as well => at the end this is the backup for operator restarts.

That's true. I'll try to see if can filter out pending changes using the ReplicasChanged condition and get rid of the pendingChanges cache, and only keep ongoing as an optimization to avoid calling Kube API every 10 seconds when polling for status.

Reviewer: It is obviously subjective, but I think you overestimate how much will this be used. The replication factor changes are super painful because of how complicated the UX of the Kafka tool is (I guess the fact nobody fixed the UX also shows how often it is used). But are IMHO not done very often. If you believe the RF to be changed all the time up and down, I also wonder how will you batch the requests (will there be a time window? Will there be a size limit?) and deal with the memory consumption of your cache.

I personally would go for simplicity, but it is your proposal. So if you wanna stick with it, just make sure it is properly described. 

> That's true. I'll try to see if can filter out pending changes using the ReplicasChanged condition, getting rid of the pendingChanges cache, and only keep the ongoing cache as an optimization to avoid calling the Kube API every 10 seconds when polling for status.

I do not understand why would you call Kubernetes API every 10 seconds. But you should definitely not do that. As I said before, you should add an explanation how will this work.

Author: > I do not understand why would you call Kubernetes API every 10 seconds.

In fact, I don't do it.

Author: I'm making few changes to make the ReplicasChangeManager completely driven by the regular reconciliation. Then, I'll update the proposal.

![Replication Factor Change](./images/063-topic-replication-factor-change.png)

The KafkaTopic resources with replicas change are added to a pendingChanges cache (Java List) the first time they are notified to the ReplicasChangeManager.
When a replicas change is marked as pending, other replicas changes for the same KafkaTopic are ignored until the pending replicas change is sent to Cruise Control and the execution is completed.

> The KafkaTopic resources marked as pending are added to an ongoingChanges cache (Java Map keyed by `User-Task-ID`) when the topic configuration request is accepted by Cruise Control.

Reviewer: Where do yu get the Task Id when it is not yet given to you by CC?

Author: A pending KT are added to the ongoing map only when we have the User-Task-ID (i.e. CC has received the request, queued the task, and replied).

In case of Cluster Operator deployment, a dedicated admin user is automatically created for the Topic Operator.

In case of standalone deployment, the Cluster Operator expects the following secrets to be created by the user:

- A secret with `cruise-control.p12` key containing the truststore in PKCS12 format, and `cruise-control.password` key containing the truststore password.
- A secret with `cruise-control.apiToAdminUsername` key containing the API username with admin role, and `cruise-control.apiToAdminPassword` key containing the user password.

#### Error handling

During the whole replicas change process, the KafkaTopic resource remains in a Ready state.
Any KafkaTopic with an ongoing replicas change has a status condition of type Ongoing and reason ReplicasChanged, and a new status field called  `userTaskId`, which is used to rebuild the cache on ReplicasChangeManager's initialization.
Pending replicas changes that were not marked as ongoing (unknown to Cruise Control), are simply re-added to pendingChanges by the Topic Operator reconciliation logic.

> Ongoing replicas change status example:

Reviewer: This is fine as an example. But I think you should have a separate section dedicated to API (CRD) changes.

Author: Ok.

```sh
status:
  conditions:
  - lastTransitionTime: "2024-01-12T14:54:02.833067402Z"
    status: "True"
    type: Ready
  - lastTransitionTime: "2024-01-12T15:01:24.395861299Z"
    message: Replicas change ongoing
    reason: ReplicasChanged
    status: "True"

>     type: Ongoing

Reviewer: Maybe some `type: Information` would be better? Or `Warning`, but this is not really a warning, this is just FYI basically.

  observedGeneration: 7
  topicName: my-topic

>   userTaskId: 750dfc95-c67b-431c-979b-43843ac177e7

Reviewer: Should there be some subsection? Something like

```
status:
  # ...
  replicationFactorChange:
    userTaskId: ...
```

This would provide more clarity to what it is and make it easier to add more fields easily if needed. (As I said, there should probably be a separate section for API/CRD changes to discuss this).

Author: Ok, thanks.

The Cluster Operator is already integrated with Cruise Control, so we can factor out the existing logic and classes.

Cruise Control creates a task for each operation, such as topic configuration and cluster rebalance.
A task is first created in the Active state, then moves to InExecution, and finally to Completed or CompletedWithError.
At most `max.active.user.tasks` tasks can be Active at the same time (5 by default), and only one task can be InExecution at any given time.
During the execution, a task is divided in one or more batches, which are then executed one after the other.

### Topic configuration

The [`topic_configuration`](https://github.com/linkedin/cruise-control/wiki/REST-APIs#change-kafka-topic-configuration) endpoint is used to request a topic configuration update, and currently only supports replication factor changes.
Once accepted, the replication factor change is put into the Cruise Control's task queue and executed asynchronously.

If partition's current replication factor is less than target replication factor, new replicas are added to the partition in two steps.

> Tentatively add new replicas in a rack-aware, round-robin way, and then further optimize location of new replicas with provided or default Goal list (same heuristic of the rebalance).

Author: This is under topic configuration section, which is a Cruise Control's endpoint, so it describes what Cruise Control does. I'm getting rid of it to avoid confusion.

> Rack awareness property is guaranteed only when the rack information is in metadata, and target replication factor is less than number or racks.

Author: Good question. This is the internal Cruise Control logic, and I don't get any feedback on that from the `topic_configuration` response.

If the current replication factor of partition is larger than target replication factor, one or more follower replicas are removed from the partition.
Replicas are removed following the reverse order of position in replica list of partition.

The two required parameters for this endpoint are `topic` (regular expression to specify subject topics) and `replication_factor` (target replication factor).

> In order to batch multiple changes in the same request, the Topic Operator uses a JSON payload containing topics grouped by their replication factor.

Author: Thanks.

> POST /kafkacruisecontrol/topic_configuration?skip_rack_awareness_check=true&dryrun=false&json=true HTTP/1.1
> Host: localhost:9090
> User-Agent: curl/8.0.1
> Accept: */*
> 
} [91 bytes data]
< HTTP/1.1 200 OK
< Date: Thu, 07 Dec 2023 14:58:55 GMT
< Set-Cookie: JSESSIONID=node0lfkx183ijq4pjkkjahsegfmp30.node0; Path=/
< Expires: Thu, 01 Jan 1970 00:00:00 GMT
< User-Task-ID: 5344ca89-351f-4565-8d0f-9aade00e053d
< Content-Type: application/json;charset=utf-8
< Cruise-Control-Version: 2.5.77-SNAPSHOT
< Cruise-Control-Commit_Id: 535ea07e8847990cc02dde1f1be99b387dbeaf5b
< Content-Length: 17051
< Server: Jetty(9.4.52.v20230823)
< 
{ [17321 bytes data]
{
  "summary": {
    "numIntraBrokerReplicaMovements": 0,
    "numReplicaMovements": 125,
    "onDemandBalancednessScoreAfter": 100,
    "intraBrokerDataToMoveMB": 0,
    "monitoredPartitionsPercentage": 100,
    "provisionRecommendation": "",
    "excludedBrokersForReplicaMove": [],
    "excludedBrokersForLeadership": [],
    "provisionStatus": "RIGHT_SIZED",
    "onDemandBalancednessScoreBefore": 100,
    "recentWindows": 5,
    "dataToMoveMB": 0,
    "excludedTopics": [],
    "numLeaderMovements": 0
  }
  ...
```

> The `skip_rack_awareness_check` parameter configures whether to skip the rack awareness sanity check or not (disabled by default).

Reviewer: Just a nit ... The way I understand it, by default the `skip_rack_awareness_check` is set to false? Maybe this would make it more clear?

```suggestion
The `skip_rack_awareness_check` parameter configures whether to skip the rack awareness sanity check or not (defaults to `false`).
```

As it is a bit confusing if the rack awareness check is disabled by default or the _skipping_ of the rack awareness check is disabled by default.

Author: Agreed.

This is configurable for all endpoints in `Kafka.spec.cruiseControl.config` using the `default.goals` property.

The `replication_throttle` parameter is an upper bound on the bandwidth used to move replicas (bytes per second).
This is configurable for all endpoints in `Kafka.spec.cruiseControl.config` using the `default.replication.throttle` property.

### User tasks

The [`user_tasks`](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-user-request-result) endpoint is used to periodically check the replicas change tasks state.
The only required parameter is the `User-Task-ID`, which is returned by the `topic_configuration` response.

```sh
$ curl -s "localhost:9090/kafkacruisecontrol/user_tasks?user_task_ids=5344ca89-351f-4565-8d0f-9aade00e053d&json=true" | jq
{
  "userTasks": [
    {
      "Status": "Completed",
      "ClientIdentity": "127.0.0.1",
      "RequestURL": "POST /kafkacruisecontrol/topic_configuration?topic=my-topic&replication_factor=2&skip_rack_awareness_check=true&dryrun=false&json=true",
      "UserTaskId": "5344ca89-351f-4565-8d0f-9aade00e053d",
      "StartMs": "1701961135978"
    }
  ],
  "version": 1
}
```

> 

Author: That's another good question, and I don't know the answer right now, but I'll try to figure it out. Added a note about tasks completed as a single unit.

### Configuration parameters

The Topic Operator needs to authenticate and send HTTP requests to the Cruise Control's REST API.

> The following environment variables are set by the Cluster Operator when Cruise Control is configured in Kafka resource, or manually by the user in case of standalone deployment:

Author: Thanks, I clearly need to build more experience with that.

| Name                                   | Type    | Default | Description                                                   |
|:---------------------------------------|---------|---------|---------------------------------------------------------------|
| STRIMZI_CRUISE_CONTROL_ENABLED         | Boolean | false   | Whether Cruise Control integration is enabled                 |

> | STRIMZI_CRUISE_CONTROL_NAMESPACE       | String  | ""      | Namespace where Cruise Control is running                     |

Author: You are correct. We don't need it.

| STRIMZI_CRUISE_CONTROL_HOSTNAME        | String  | ""      | Cruise Control hostname                                       |
| STRIMZI_CRUISE_CONTROL_PORT            | Integer | 9090    | Cruise Control port                                           |
| STRIMZI_CRUISE_CONTROL_SSL_ENABLED     | Boolean | false   | Whether SSL encryption is enabled                             |
| STRIMZI_CRUISE_CONTROL_AUTH_ENABLED    | Boolean | false   | Whether Basic authentication is enabled                       |
| STRIMZI_CRUISE_CONTROL_RACK_ENABLED    | Boolean | false   | Whether rack awareness is enabled in the target Kafka cluster |
| STRIMZI_CRUISE_CONTROL_SECRET_NAME     | String  | ""      | Secret name containing the truststore                         |
| STRIMZI_CRUISE_CONTROL_API_SECRET_NAME | String  | ""      | Secret name containing the API credentials                    |

With Cluster Operator deployment, a dedicated admin user is automatically created for the Topic Operator.
With standalone deployment, the Cluster Operator expects two secrets that needs to be created by the user:

> - A secret with `cruise-control.p12` key containing the truststore in PKCS12 format, and `cruise-control.password` key containing the truststore password.

Reviewer: Should this use PEM file(s) instead? I think in general our interface is usually based on PEM files and the PKCS12 files are created internally only.

Author: Will look at that. Thanks.

- A secret with `cruise-control.apiToAdminUsername` key containing the API username with admin role, and `cruise-control.apiToAdminPassword` key containing the user password.

### KafkaTopic CRD changes

During the whole replicas change process the KafkaTopic remains in a Ready state.
A replicas change is executed asynchronously, and takes more than one reconciliation to complete.

A new optional KafkaTopicStatus subsection called `replicasChange` is used to update the user, and keep track of the replicas change state.
At any given time, the replicas change can be pending, ongoing, or completed.

- **Pending**: Not in Cruise Control's task queue (not yet sent or request error).
  ```sh
  status:
    conditions:
    - lastTransitionTime: "2024-01-18T16:13:50.490918232Z"
      status: "True"
      type: Ready
    observedGeneration: 2
    replicasChange:

>       completed: false

Reviewer: Not a blocker. But does it make sense to have a `completed` field that seems to have only one option of `false`? Or when it would be `true`? Would it make more sense to track the state there? E.g. `state: Pending`, state: Ongoing` etc.? It might give you more flexibility if in the future you need to do something more in these states as well.

Reviewer 2: I agree that it's better to record the resource states directly, rather than just the boolean, and also describe how these states map to the CC states (they may or may not be the same, depending how how things work).

Author: This is a good idea and something that I also tried at some point in my draft implementation and lost while working on status updates. I'm adding it back. 

The state will show pending and then ongoing. It won't show the completed state to avoid confusion. I will simply remove the entire `replicasChange` subsection when done.

    topicName: my-topic
  ```

- **Ongoing**: In Cruise Control's task queue, but execution not started, or not completed.
  ```sh
  status:
    conditions:
    - lastTransitionTime: "2024-01-18T16:13:53.490918232Z"
      status: "True"
      type: Ready
    observedGeneration: 3
    replicasChange:
      completed: false

>       userTaskId: 1aa418ca-53ed-4b93-b0a4-58413c4fc0cb

Author: Looks like we are using `sessionId` in the `KafkaRebalanceStatus`, would that work?

    topicName: my-topic
  ```

- **Completed**: Cruise Control's task execution completed (target replication factor reconciled).
  ```sh
  status:
    conditions:
    - lastTransitionTime: "2024-01-18T16:13:58.490918232Z"
      status: "True"
      type: Ready
    observedGeneration: 4
    topicName: my-topic
  ```

### Reconciliation logic

At runtime, the Topic Operator watches and periodically reconciles the configuration of all managed and unpaused KafkaTopic resources.
Among the other configurations, the reconciliation logic detects `KafkaTopic.spec.replicas` changes by comparing with the topic replication factor value in Kafka.

A ReplicasChangeClient is used to isolate the logic required to call the Cruise Control REST API from the rest of the reconciliation logic.

> The ReplicasChangeClient supports sending multiple replicas changes at the same time.

Reviewer: Does this mean batching multiple topics into a single request? Or just having separate requests for different topics sent in parallel?

Reviewer 2: I agree with need a bit more clarity on this. I took it to mean multiple topics get batched into a single request. But we need to understand _how_ that happens. Imagine the following scenario:

* User has some script/tooling to update `KafkaTopic` RF in bulk.
* The script does a `kubteclt patch` on the `KafkaTopics` in a loop.
* The TO reacts to those watch events one at a time
* What we want to happen: We hit the `topic_configuration` endpoint once, with all those topics in the request.
* What the easy implementation does: We hit the `topic_configuration` endpoint once for each topic.

We want the former behaviour because it's reduces the amount of time the user has to wait for their changes to take effect, and because it means CC can do a single set of data movements, avoiding unnecessary work. But it's not completely trivial to get this behaviour. One way to do it would be to have a queue of RF changes. The regular reconciliation logic adds to the queue, and we process this queue in batches. But we'd need to define it in more detail how this happens, and what configuration parameters ought to be exposed.

Reviewer: I disagree. I think we want to go for the simplicity of the implementation over having an amazing efficiency in niche use-cases. (Also ... LOL ... we have been discussing this for several weeks and Federico just changed this :-o)

Author: I switched to batch/group requests, and it's not more complicated, really. I just have to group the topics by RF value and create the json. There is no additional queue, the reconciliation logic already has one, and provides a batch of replicas changes. 

This is useful with bulk replicas change, which can happen in some edge cases. A potentially big backlog of messages may be created when the Topic Operator is not running, or there is an issue with Cruise Control or Kafka. Another example is a GitOps pipeline delivering many replicas changes at the same time to reduce the storage occupation of non-critical topics.

I tried my draft implementation with 100 replicas updates applied at the same time (not in a loop) on topics with some data. This created 2 or 3 Cruise Control's requests and completed in less that 5 minutes. I think this is good enough behavior.

Reviewer: I guess this is not obvious from the proposal -> neither that it works like that nor how it is actually implemented.

Author: Yeah, I realized that and pushed some changes in the "Topic configuration" section.

On every reconciliation loop, when Cruise Control integration is enabled and replicas changes are detected, the BatchingTopicController uses the ReplicasChangeClient for the following operations:

1. Request pending replicas changes (uses `topic_configuration` endpoint, and returns topics with pending and ongoing changes).

> 2. Check ongoing replicas changes (uses `user_tasks` endpoint, and returns topics with completed changes).

Reviewer: Does this query all existing tasks? Or how does it find out what topics are being processed right now? I think it is the `userTaskId` that you have to query.

Author: It queries all existing tasks with one request. You basically pass a comma separated list of `userTaskId`s, as described in the `user_tasks` endpoint section of this proposal.

The result of these operations is used to update the KafkaTopic status.

### Error handling

> When there is an error, the replicas change remains pending or ongoing (no status update), and it is indefinitely retried by the periodic reconciliation.

Reviewer: Should there be some backoff for example similar to the connector auto-restart?

Author: We don't have anything similar for the other topic configurations, the Topic Operator simply retries indefinitely during periodic reconciliations. As soon as the problem is fixed, the change will be reconciled. We can always add it later if there is demand, but I doubt.

> All errors are written in the Topic Operator logs, so that the user can verify the reason of a never ending replicas change.

Reviewer: Should/can the errors be added also to the conditions as warnings? I'm not sure if they are persistent, so not sure if it is really possible or if they will be added and immediately removed in the next reconciliation few milliseconds later.

Author: Added a message field to `replicasChange`, that we can use for any message to the user. For the moment I'll use it only for errors/warnings.

In Cruise Control, there is no way to differentiate between temporary errors (resolve by themself) from permanent errors (require some action).
For example, the `NotEnoughValidWindowsException` can be raised when Cruise Control has just started and the cluster model is still building, but also when there is a configuration error which prevents broker metrics collection.
Another example is the `OptimizationFailureException` raised when some hard goal is violated, which can be temporary if the configured network capacity is less than the real network capacity and there is a traffic peak.
Instead, the `OngoingExecutionException` is clearly temporary and it means that there is already a task that is being executed by Cruise Control.

Cruise Control allows to scale down the replication factor under the `min.insync.replicas` value, and this can cause disruption to producers with `acks=all`.

> When this happens, the Topic Operator logs a warning but continues with the replicas change, because the KafkaRoller ignores topics with RF < minISR, and they don't event show up as under replicated.

Reviewer 2: ```suggestion
When this happens, the Topic Operator logs a warning but continues with the replicas change, because the KafkaRoller ignores topics with RF < minISR, and they don't even show up as under replicated.
```

Don't show up as under-replicated _where_?

Author: Metrics and tools. Fixed. Thanks.

> The target replication factor should also be less than or equal to the number of available brokers, but this is enforced directly by Cruise Control.

Reviewer 2: Meaning this manifests to the TO as an error on the request to be topic_configration endpoint?

Author: Exactly.

> When a manged KafkaTopic with replicas change is deleted in Kubernetes, the Topic Operator also deletes the topic in Kafka, but the task execution continues in Cruise Control.

Reviewer: ```suggestion
When a managed KafkaTopic with replicas change is deleted in Kubernetes, the Topic Operator also deletes the topic in Kafka, but the task execution continues in Cruise Control.
```

Author: Thanks.

> There is little benefit in using the `stop_proposal_execution` endpoint, because most replicas change tasks will have only one batch that can't be stopped.

Reviewer: So, what happens with such task? Will it just error in CC? I guess it is important to make sure it does not keep hanging there because of some race condition. Should we use the finalizers to prevent this from happening and delete the topic only once the task completes?

(to be clear, I'm fine with what you propose and not adding the complexity if we know for sure it will not have any negative side-effects)

Author: CC simply continues to execute the batch to the end. It's unnecessary work, because the topic is already deleted, but there is no harm (no task hang or error). Clarified in the proposal.

When a manged topic with replicas change is deleted in Kafka, the Topic Operator recreates the topic with the target replication factor.

When Topic Operator is restarted, both pending and ongoing replicas changes are recovered from KafkaTopic status.
The ongoing state includes the `userTaskId`, which is the unique identifier of a Cruise Control task, which includes one or more replicas changes.

When Cruise Control is restarted, all active tasks are lost, as there is no persistent memory.
The Topic Operator detects this event when the ongoing replicas changes check returns no tasks for a known `userTaskId`.

> All pending and ongoing replicas changes are still stored in KafkaTopic status, and can be recovered as soon as Cruise Control becomes Ready.

Reviewer: This seems weird. This suggests some kind of recovery of where you were before. But from the KafkaTopic CR changes it seems there is no way to do it -> so I guess you will just create a new task depending on / corresponding to the current state?

Reviewer 2: I think the TO will have to make another request to the topic_configuration endpoint in this case. But I agree that the proposal should say this explicitly.

Author: Yes, new tasks are created for all changes that were not applied when CC restarted. Clarified in the proposal.

> 

Reviewer 2: You said before that CC allows only a fixed number of pending tasks, so I'd like to see some discussion of how the TO handles that situation, including show it's reflected in the `KafkaTopic`.

Author: Sure, I added that part. It's the same error handling procedure. A clear message is surfaced to the user (There are already 5 active user tasks, which has reached the servlet capacity.), and the Topic Operator periodically retries until there is room for it.
